File: removeNthFromEnd.java
/**
 * Definition for singly-linked list.
 * public class ListNode {
 *     int val;
 *     ListNode next;
 *     ListNode() {}
 *     ListNode(int val) { this.val = val; }
 *     ListNode(int val, ListNode next) { this.val = val; this.next = next; }
 * }
 */
/* 
class Solution {
    public ListNode removeNthFromEnd(ListNode head, int n) {
        //Tc: O(N)  Sc: O(1)
    ListNode dummy = new ListNode(0);
    dummy.next = head;
    ListNode p1 = dummy;
    ListNode p2 = dummy;

    for (int i = 1; i <= n + 1; i++) {
        p1 = p1.next;
    }

    while (p1 != null) {
        p1 = p1.next;
        p2 = p2.next;
    }
    p2.next = p2.next.next;
    return dummy.next;
        
    }
}

*/

File: removeNthFromEnd.js
// Time Complexity : O(n)
// Space Complexity : O(1)
// Did this code successfully run on Leetcode : Yes
// Any problem you faced while coding this : No

/**
 * Definition for singly-linked list.
 * function ListNode(val, next) {
 *     this.val = (val===undefined ? 0 : val)
 *     this.next = (next===undefined ? null : next)
 * }
 */
/**
 * @param {ListNode} head
 * @param {number} n
 * @return {ListNode}
 */

var removeNthFromEnd = function(head, n) {
    let dummy = new ListNode();
    let fast = dummy;
    let slow = dummy;
    dummy.next = head;
    let count = 0;
    while(count < n) {
        fast = fast.next;
        count++;
    }
    while(fast.next != null) {
        slow = slow.next;
        fast = fast.next;
    }
    slow.next = slow.next.next;
    return dummy.next;
};